pkg/ingress/gateway.go

```diff
@@ -134,58 +77,80 @@ func (c *client) handleCertificateChange(currentCertSpec *configv1alpha2.Ingress
-			}
+func shouldRelisten(oldSpec *configv1alpha2.IngressGatewayCertSpec, newSpec *configv1alpha2.IngressGatewayCertSpec) bool {
+	if newSpec == nil || len(newSpec.SubjectAltNames) == 0 {
+		return false
+	}
 
-			// This should never happen, but guarding against a panic due to the usage of a pointer
-			if currentCertSpec == nil {
-				log.Error().Msgf("Current ingress gateway cert spec is nil, but a certificate for it was rotated - unexpected")
-				continue
-			}
+	if oldSpec == nil || len(oldSpec.SubjectAltNames) == 0 {
+		return true
+	}
 
-			cnInCertSpec := currentCertSpec.SubjectAltNames[0] // Only single SAN is supported in certs
+	return oldSpec.SubjectAltNames[0] != newSpec.SubjectAltNames[0]
+}
 
-			// Only update the secret if the cert rotated matches the cert spec
-			if cert.GetCommonName() != certificate.CommonName(cnInCertSpec) {
-				continue
-			}
+// watchCertificateRotations creates and stores a certificate with the given common name into the given secret ref.
+// It then watches for all cert rotations and updates the secret on changes.
+func (c *client) watchCertificateRotations(cn string, secret corev1.SecretReference) func() {
+	// This is some fanciness to provide the following guarantees:
+	// 1. The subscription is called before a goroutine starts, allowing the subscription creation to be deterministic.
+	// 2. By signalling to stop (instead of just closing it), we guarantee the close func doesn't return before the goroutine stops.
+	// 3. The once allows the close method to be called multiple times.
+	var once sync.Once
+	stop := make(chan struct{})
 
-			log.Info().Msg("Ingress gateway certificate was rotated, updating corresponding secret")
-			if err := c.createAndStoreGatewayCert(*currentCertSpec); err != nil {
-				log.Error().Err(err).Msgf("Error updating ingress gateway cert secret after cert rotation")
-			}
+	// must subscribe prior to issuing the cert to guarantee we get all rotations.
+	certRotateChan, unsub := c.certProvider.SubscribeRotations(cn)
 
-		case <-stop:
-			return
+	cert, err := c.certProvider.IssueCertificate(cn, certificate.IngressGateway, certificate.FullCNProvided())
+	if err != nil {
+		log.Err(err).Msg("error issuing a certificate for ingress gateway")
+	}
+	if err := c.storeCertInSecret(cert, secret); err != nil {
+		log.Err(err).Msg("Error updating ingress gateway cert secret after cert rotation")
+	}
+
+	go func() {
+		for {
+			select {
+			// A certificate was rotated
+			case msg := <-certRotateChan:
+				cert := msg.(*certificate.Certificate)
+				log.Info().Msg("Ingress gateway certificate was rotated, updating corresponding secret")
+				if err := c.storeCertInSecret(cert, secret); err != nil {
+					log.Err(err).Msg("Error updating ingress gateway cert secret after cert rotation")
+				}
+
+			case <-stop:
+				return
+			}
 		}
+	}()
+	return func() {
+		// Allow close to be called multiple times by wrapping it in a once.
+		once.Do(func() {
+			stop <- struct{}{}
```

Comment:
Shouldn't `close(stop)` be sufficient to call it done?

Reply:
close is non-blocking, which could allow race conditions. I'll add a comment explaining more